Prefer element type conversions over collection type conversion comparisons (#8377)
dotnet/csharplang: proposals/collection-expressions-better-conversion.md

Given:
- `E` is a collection expression with element expressions `[EL₁, EL₂, ..., ELₙ]`
- `T₁` and `T₂` are collection types
- `E₁` is the element type of `T₁`
- `E₂` is the element type of `T₂`
- `CE₁ᵢ` are the series of conversions from `ELᵢ` to `E₁`
- `CE₂ᵢ` are the series of conversions from `ELᵢ` to `E₂`

> If `E₁` and `E₂` have an identity conversion, then the element conversions are as good as each other. Otherwise, the element conversions to `E₁` are better than the element conversions to `E₂` if:

Reviewer: ```suggestion
If there is an identity conversoin from `E₁` to `E₂`, then the element conversions are as good as each other. Otherwise, the element conversions to `E₁` are better than the element conversions to `E₂` if:
```

> If there is an identity conversoin from `E₁` to `E₂`, then the element conversions are as good as each other. Otherwise, the element conversions to `E₁` are better than the element conversions to `E₂` if:

Reviewer 2: ```suggestion
If there is an identity conversion from `E₁` to `E₂`, then the element conversions are as good as each other. Otherwise, the element conversions to `E₁` are better than the element conversions to `E₂` if:
```

- For every `ELᵢ`, `CE₁ᵢ` is at least as good as `CE₂ᵢ`, and
- There is at least one i where `CE₁ᵢ` is better than `CE₂ᵢ`
Otherwise, neither set of element conversions is better than the other, and they are also not as good as each other.  
Conversion comparisons are made using better conversion from expression if `ELᵢ` is not a spread element. If `ELᵢ` is a spread element, we use better conversion from the element type of the spread collection to `E₁` or `E₂`, respectively.

> `T₁` is a ***better collection conversion from expression*** than `T₂` if:

Reviewer: ```suggestion
`C₁` is a ***better collection conversion from expression*** than `C₂` if:
```

> - `T₁` is implicitly convertible to `T₂`, and `T₂` is not implicitly convertible to `T₁`, or

Reviewer: I think we'll need to rule out span conversions here.
```suggestion
- `T₁` or `T₂` is not a *span type*, and `T₁` is implicitly convertible to `T₂`, and `T₂` is not implicitly convertible to `T₁`, or
```

| `ReadOnlySpan<object>` | `Span<string>` | `[new object()]` | `[Identity]` | Not applicable | `T₁` is not applicable | `ReadOnlySpan<object>` is picked |
| `ReadOnlySpan<InterpolatedStringHandler>` | `ReadOnlySpan<string>` | `[${1}]` | `[Interpolated String Handler]` | `[Identity]` | `CE₁ᵢ` is better | `ReadOnlySpan<InterpolatedStringHandler>` is picked |
| `ReadOnlySpan<InterpolatedStringHandler>` | `ReadOnlySpan<string>` | `[${"blah"}]` | `[Interpolated String Handler]` | `[Identity]` - But constant | `CE₂ᵢ` is better | `ReadOnlySpan<string>` is picked |
| `ReadOnlySpan<string>` | `ReadOnlySpan<FormattableString>` | `[${1}]` | `[Identity]` | `[Interpolated String]` | `CE₂ᵢ` is better | `ReadOnlySpan<string>` is picked |
| `ReadOnlySpan<string>` | `ReadOnlySpan<FormattableString>` | `[${1}, (FormattableString)null]` | Not applicable | `[Interpolated String, Identity]` | `T₁` isn't applicable | `ReadOnlySpan<FormattableString>` is picked |
| `HashSet<short>` | `Span<long>` | `[1, 2]` | `[Implicit Constant, Implicit Constant]` | `[Implicit Numeric, Implicit Numeric]` | `CE₁ᵢ` is better | `HashSet<short>` is picked |

> | `HashSet<long>` | `Span<short>` | `[1, 2]` | `[Implicit Numeric, Implicit Numeric]` | `[Implicit Constant, Implicit Constant]` | `CE₂ᵢ` is better | `Span<short>` is picked |

Reviewer 3: not reviewing this.

## Open questions

### How far should we prioritize `ReadOnlySpan`/`Span` over other types?

As specified today, the following overloads would be ambiguous:

```cs
C.M1(["Hello world"]); // Ambiguous, no tiebreak between ROS and List
C.M2(["Hello world"]); // Ambiguous, no tiebreak between Span and List

C.M3(["Hello world"]); // Ambiguous, no tiebreak between ROS and MyList.

C.M4(["Hello", "Hello"]); // Ambiguous, no tiebreak between ROS and HashSet. Created collections have different contents

class C
{
    public static void M1(ReadOnlySpan<string> ros) {}
    public static void M1(List<string> list) {}

    public static void M2(Span<string> ros) {}
    public static void M2(List<string> list) {}

    public static void M3(ReadOnlySpan<string> ros) {}
    public static void M3(MyList<string> list) {}

    public static void M4(ReadOnlySpan<string> ros) {}

>     public static void M4(HashSet<string> hashset) {}

Reviewer: Consider removing the extra examples and using `M1` and `M2`. (Are any changes to the rules likely to affect `M3` or `M4` differently?)

Author: Yes, I think they are potentially going to affect them differently. I included all the examples I think the LDM needs to see to make a decision.
